src/include/denc.h

```diff
@@ -1590,7 +1590,7 @@ inline std::enable_if_t<traits::supported && !traits::need_contiguous> decode(
     t.copy_shallow(remaining, tmp);
     auto cp = std::cbegin(tmp);
     traits::decode(o, cp);
-    p.advance(cp.get_offset());
+    p += cp.get_offset();
```

Comment:
Yeah, this looks much more natural. It might be worth asking GCC to construct the inventory of all `advance()` users:

```
commit 3fa5d07fa6ca8b01049858837ebabef45fec9b05
Author: Radoslaw Zarzynski <rzarzyns@redhat.com>
Date:   Mon Jan 20 13:11:42 2020 +0100

    DEBUG, bl: mark copy and copy_in of bl as deprecated.
    
    Signed-off-by: Radoslaw Zarzynski <rzarzyns@redhat.com>

diff --git a/src/include/buffer.h b/src/include/buffer.h
index dcf6602..db012f5 100644
--- a/src/include/buffer.h
+++ b/src/include/buffer.h
@@ -1152,11 +1152,11 @@ inline namespace v14_2_0 {
 
     // crope lookalikes.
     // **** WARNING: this are horribly inefficient for large bufferlists. ****
-    void copy(unsigned off, unsigned len, char *dest) const;
-    void copy(unsigned off, unsigned len, list &dest) const;
-    void copy(unsigned off, unsigned len, std::string& dest) const;
-    void copy_in(unsigned off, unsigned len, const char *src, bool crc_reset = true);
-    void copy_in(unsigned off, unsigned len, const list& src);
+    void copy(unsigned off, unsigned len, char *dest) const __attribute__((deprecated("BL HUNT: copyE")));
+    void copy(unsigned off, unsigned len, list &dest) const __attribute__((deprecated("BL HUNT: copyE")));
+    void copy(unsigned off, unsigned len, std::string& dest) const __attribute__((deprecated("BL HUNT: copyE")));
+    void copy_in(unsigned off, unsigned len, const char *src, bool crc_reset = true) __attribute__((deprecated("BL HUNT: copy_inE")));
+    void copy_in(unsigned off, unsigned len, const list& src) __attribute__((deprecated("BL HUNT: copy_inE")));
```

I can do that on the occasion of total repo rebuild I already need to do. Maybe we could knock `advance()` out entirely.

Reply:
yeah, i am for this, will add a commit to drop `advance()`, as it's a C++ API. IIRC, we don't keep the backward compatibility of C++ ABI/API.

Reply:
> IIRC, we don't keep the backward compatibility of C++ ABI/API

That's also my understanding. All we need to do is to ensure nobody will link with the changed stuff by accident – and that can be accomplished by just bumping up the version in our symbols' names.

Reply:
Related comment: https://github.com/ceph/ceph/pull/32806#discussion_r370200811.

Reply:
@rzarzynski updated and repushed.

Reply:
Just to the record:

```
$ grep "BL HUNT: bl_advanceE" /tmp/20200131_bl_multihunt.txt |  cut -d: -f 1,2 | sed -e 's#/work/ceph-1/##' | sort | uniq
src/common/buffer.cc:672
src/common/buffer.cc:708
src/common/buffer.cc:725
src/common/buffer.cc:762
src/common/buffer.cc:798
src/common/buffer.cc:817
src/common/buffer.cc:837
src/common/buffer.cc:854
src/common/buffer.cc:942
src/compressor/snappy/SnappyCompressor.h:49
src/include/denc.h:1593
src/include/denc.h:1614
src/include/denc.h:1653
src/include/encoding.h:1390
src/include/encoding.h:1475
src/journal/Entry.cc:105
src/journal/Entry.cc:118
src/mds/journal.cc:3094
src/messages/MOSDPing.h:113
src/msg/async/ProtocolV1.cc:865
src/objclass/../include/denc.h:1593
src/objclass/../include/denc.h:1614
src/objclass/../include/encoding.h:1475
src/os/bluestore/bluestore_types.cc:39
src/osd/OSDMap.h:145
src/osd/osd_types.cc:317
src/osd/osd_types.cc:5579
src/osd/osd_types.h:481
src/osd/PrimaryLogPG.cc:5202
src/os/filestore/FileJournal.h:187
src/os/Transaction.cc:17
src/os/Transaction.cc:20
src/os/Transaction.cc:36
src/rgw/services/svc_notify.cc:399
src/rgw/services/svc_notify.cc:438
src/test/bufferlist.cc:1055
src/test/bufferlist.cc:1095
src/test/bufferlist.cc:1116
src/test/bufferlist.cc:1132
src/test/bufferlist.cc:1158
src/test/bufferlist.cc:1175
src/test/bufferlist.cc:797
src/test/bufferlist.cc:826
src/test/bufferlist.cc:831
src/test/bufferlist.cc:833
src/test/bufferlist.cc:932
src/test/bufferlist.cc:939
src/test/bufferlist.cc:963
src/test/compressor/test_compression.cc:204
src/test/journal/test_Entry.cc:73
src/test/test_denc.cc:715
src/test/test_denc.cc:733
```

Reply:
@rzarzynski thanks for the list. all occurrences should have been fixed now.